mlx/fast.cpp

```diff
@@ -859,4 +865,214 @@ array affine_dequantize(
+  }
+  if (output_shapes_.size() != output_dtypes_.size()) {
+    validated = false;
+  } else {
+    for (auto kv : output_shapes_) {
+      if (output_dtypes_.find(kv.first) == output_dtypes_.end()) {
+        validated = false;
+      }
+    }
+  }
+  if (!validated) {
+    throw std::invalid_argument(
+        "`output_shapes` and `output_dtypes` must have the same keys.");
+  }
+}
+
+void MetalKernel::write_signature(
+    std::string& func_name,
+    std::map<std::string, array>& inputs,
+    std::ostringstream& kernel_source) {
+  // Auto-generate a function signature based on `template_args`
+  // and the dtype/shape of the arrays passed as `inputs`.
+  if (template_args.size() > 0) {
+    kernel_source << "template <";
+    int i = 0;
+    for (auto [name, arg] : template_args) {
+      std::string param_type;
+      if (std::holds_alternative<int>(arg)) {
+        param_type = "int";
+      } else if (std::holds_alternative<bool>(arg)) {
+        param_type = "bool";
+      } else if (std::holds_alternative<Dtype>(arg)) {
+        param_type = "typename";
+      }
+      if (i > 0) {
+        kernel_source << ", ";
+      }
+      kernel_source << param_type << " " << name;
+      i++;
+    }
+    kernel_source << ">" << std::endl;
+  }
+  kernel_source << "[[kernel]] void " << func_name << "(" << std::endl;
+
+  // Metal attributes are automatically added to the arguments if present
+  const std::vector<std::pair<std::string, std::string>> metal_attributes = {
+      {"dispatch_quadgroups_per_threadgroup", "uint"},
+      {"dispatch_simdgroups_per_threadgroup", "uint"},
+      {"dispatch_threads_per_threadgroup", "uint3"},
+      {"grid_origin", "uint3"},
+      {"grid_size", "uint3"},
+      {"quadgroup_index_in_threadgroup", "uint"},
+      {"quadgroups_per_threadgroup", "uint"},
+      {"simdgroup_index_in_threadgroup", "uint"},
+      {"simdgroups_per_threadgroup", "uint"},
+      {"thread_execution_width", "uint"},
+      {"thread_index_in_quadgroup", "uint"},
+      {"thread_index_in_simdgroup", "uint"},
+      {"thread_index_in_threadgroup", "uint"},
+      {"thread_position_in_grid", "uint3"},
+      {"thread_position_in_threadgroup", "uint3"},
+      {"threadgroup_position_in_grid", "uint3"},
+      {"threadgroups_per_grid", "uint3"},
+      {"threads_per_grid", "uint3"},
+      {"threads_per_simdgroup", "uint"},
+      {"thread_per_threadgroup", "uint3"},
+  };
+  std::vector<std::pair<std::string, std::string>> attrs;
+  for (auto [attr, dtype] : metal_attributes) {
+    if (source_.find(attr) != std::string::npos) {
+      attrs.push_back({attr, dtype});
+    }
+  }
+
+  int index = 0;
+  constexpr int max_constant_array_size = 8;
+  // Add inputs
+  for (auto& [name, arr] : inputs) {
+    auto dtype = get_type_string(arr.dtype());
+    std::string location = "constant";
+    std::string ref = "*";
+    // Single ints/float/bools are passed by ref
+    if (arr.ndim() == 0) {
```

Comment:
Unfortunately that is not correct. They should be passed as constants iff they are available at dispatch time. Otherwise they have to be passed as device pointers or device references.